mlir/lib/Analysis/Presburger/Barvinok.cpp

```diff
@@ -63,3 +63,79 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
 
   return cone.determinant();
 }
+
+/// Compute the generating function for a unimodular cone.
+/// This consists of a single term of the form
+/// x^num / prod_j (1 - x^den_j)
+///
+/// den_j is defined as the set of generators of the cone.
+/// num is computed by expressing the vertex as a weighted
+/// sum of the generators, and then taking the floor of the
+/// coefficients.
+GeneratingFunction mlir::presburger::detail::unimodularConeGeneratingFunction(
+    ParamPoint vertex, int sign, ConeH cone) {
+  // `cone` must be unimodular.
+  assert(getIndex(getDual(cone)) == 1 && "input cone is not unimodular!");
+
+  unsigned numVar = cone.getNumVars();
+  unsigned numIneq = cone.getNumInequalities();
+
+  // Thus its ray matrix, U, is the inverse of the
+  // transpose of its inequality matrix, `cone`.
+  FracMatrix transp(numVar, numIneq);
+  for (unsigned i = 0; i < numVar; ++i)
+    for (unsigned j = 0; j < numIneq; ++j)
+      transp(j, i) = Fraction(cone.atIneq(i, j), 1);
+
+  FracMatrix generators(numVar, numIneq);
+  transp.determinant(/*inverse=*/&generators); // This is the U-matrix.
+
+  // The powers in the denominator of the generating
+  // function are given by the generators of the cone,
+  // i.e., the rows of the matrix U.
+  std::vector<Point> denominator(numIneq);
+  ArrayRef<Fraction> row;
+  for (unsigned i = 0; i < numVar; ++i) {
+    row = generators.getRow(i);
+    denominator[i] = Point(row);
+  }
+
+  // The vertex is v : d x (n+1)
+  // We need to find affine functions of parameters λ_i(p)
+  // such that v = Σ λ_i(p)*u_i.
+  // The λi are given by the columns of Λ = v^T U^{-1} = v^T transp.
```

Comment:
lambda_i

Reply:
`v^T transp.`?

Reply:
v^T multiplied by the matrix transp, is it not clear? How do I rewrite it?

Reply:
ok, I thought `transp.` was abbreviation of the word "transpose". It's easier to follow if the documentation names are all matched to the variable names, you could change either to match the other

Reply:
wait it does, the variable name is `transp`
should I name it something else?

Reply:
`The λi` -> `The λ_i`

Reply:
yeah I meant the other variables (matrix `U` in particular). But it's a bit hard to write these equations with long names

Reply:
I guess for now just write `Λ = v^T U^{-1}, and we have transp = U^{-1}` so it's clear

Reply:
okay